src/io.cpp

```diff
@@ -0,0 +1,567 @@
+
+    FileOpenError(std::string message, int value) : message(std::move(message)), value(value) {}
+};
+} // namespace
+
+auto type(const CallContext& ctx) -> CallResult {
+    auto file = ctx.arguments().get(0);
+
+    // return the result of calling the "type" method if the file is a table
+    // and the method exists, otherwise return nil
+    return std::visit(
+        overloaded{
+            [&ctx](const Table& table) {
+                if (table.get("type").is_function()) {
+                    return table.get("type").call(ctx);
+                } else {
+                    return CallResult();
+                }
+            },
+            [](const auto& /*unused*/) { return CallResult(); }},
+        file.raw());
+}
+
+class CFileHandle : public FileHandle {
+    // NOTE: a nullptr is treated as a closed file
+    FILE* handle;
+    bool should_close;
+
+public:
+    CFileHandle(const std::string& path, const std::string& mode)
+        : CFileHandle(std::fopen(path.c_str(), mode.c_str()), true) {}
+    CFileHandle(FILE* handle, bool should_close) : handle(handle), should_close(should_close) {
+        if (this->handle == nullptr) {
+            throw FileOpenError("could not open file", errno);
+        }
+    }
+    ~CFileHandle() override {
+        if (this->should_close && this->handle != nullptr) {
+            fclose(this->handle);
+            this->handle = nullptr;
+        }
+    }
+
+    auto is_open() -> bool override { return this->handle != nullptr; }
+    auto is_at_eof() -> bool override { return std::feof(this->handle) != 0; }
+
+    auto close() -> bool override {
+        this->ensure_file_is_open();
+
+        auto result = fclose(this->handle);
+        this->handle = nullptr;
+        return result == 0;
+    }
+    auto flush() -> bool override {
+        this->ensure_file_is_open();
+
+        return fflush(this->handle) == 0;
+    }
+    auto read_all() -> Value override {
+        // figure out how long the file is (from the current position)
+        // and reset position
+        long current_position = ftell(this->handle);
+        fseek(this->handle, 0L, SEEK_END);
+        long numbytes = ftell(this->handle) - current_position;
+        fseek(this->handle, current_position, SEEK_SET);
+
+        // allocate a buffer
+        auto* buffer = new char[numbytes];
+
+        // read the file to the buffer
+        fread(buffer, sizeof(char), numbytes, this->handle);
+
+        std::string content(buffer, numbytes);
+        delete[] buffer;
+
+        return content;
+    }
+
+    auto read_num() -> Value override {
+        const size_t READ_BUFFER_SIZE = 64;
+
+        std::string str;
+        char buffer[READ_BUFFER_SIZE];
+
+        auto current_position = ftell(this->handle);
+
+        // consume whitespace
+        fscanf(this->handle, " ");
+        // collect string that could potentially be a number
+        while (fscanf(this->handle, "%63[abcdefABCDEF0123456789.x-]", buffer) > 0) {
```

Comment:
This is not a regex pattern. According to the documentation this can work on some implementations but it is not required by the standard and therefore undefined behavior.